File: rust/services/order/app/src/core/port/outgoing.rs
use crate::core::{model::order::Order, service::order_id::OrderId};
use async_trait::async_trait;
use std::sync::Arc;

pub type DynOrderRepository = Arc<dyn OrderRepository + Send + Sync>;

#[async_trait]
pub trait OrderRepository {
    async fn find_all_orders(&self, offset: i64, limit: i64) -> Result<Vec<Order>, String>;
    async fn find_order_by_id(&self, order_id: OrderId) -> Result<Order, String>;
    async fn save_order(&self, order: Order) -> Result<Order, String>;
}

File: services/order/app/core/port/outgoing.go
package port

import (
	"context"
	"errors"
	"monorepo/services/order/app/core/model"

	"github.com/google/uuid"
)

var (
	OrderNotFoundError = errors.New("order not found")
)

type OrderRepository interface {
	FindAllOrdersByCustomerId(ctx context.Context, customerId uuid.UUID, offset int, limit int) ([]model.Order, error)
	FindAllOrders(ctx context.Context, offset int, limit int) (orders []model.Order, err error)
	FindOrderByOrderId(ctx context.Context, orderId model.OrderId) (order model.Order, err error)
	SaveOrder(ctx context.Context, order model.Order) error
}
